src/streaming/file.sink.cpp

```diff
@@ -3,31 +3,226 @@
+#endif
+
 namespace fs = std::filesystem;
 
-zarr::FileSink::FileSink(std::string_view filename, bool truncate)
-  : file_(filename.data(),
-          truncate ? (std::ios::binary | std::ios::trunc) : std::ios::binary)
+namespace {
+#ifdef _WIN32
+std::string
+get_last_error_as_string()
+{
+    DWORD errorMessageID = ::GetLastError();
+    if (errorMessageID == 0) {
+        return std::string(); // No error message has been recorded
+    }
+
+    LPSTR messageBuffer = nullptr;
+
+    size_t size = FormatMessageA(FORMAT_MESSAGE_ALLOCATE_BUFFER |
+                                   FORMAT_MESSAGE_FROM_SYSTEM |
+                                   FORMAT_MESSAGE_IGNORE_INSERTS,
+                                 NULL,
+                                 errorMessageID,
+                                 MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+                                 (LPSTR)&messageBuffer,
+                                 0,
+                                 NULL);
+
+    std::string message(messageBuffer, size);
+
+    LocalFree(messageBuffer);
+
+    return message;
+}
+
+void
+init_handle(void** handle, const std::string& filename)
+{
+    EXPECT(handle, "Expected nonnull pointer to file handle.");
+    HANDLE* fd = new HANDLE;
+
+    *fd = CreateFileA(filename.c_str(),
+                      GENERIC_WRITE,
+                      0, // No sharing
+                      nullptr,
+                      OPEN_ALWAYS,
+                      FILE_FLAG_OVERLAPPED,
+                      nullptr);
+
+    if (*fd == INVALID_HANDLE_VALUE) {
+        const auto err = get_last_error_as_string();
+        delete handle;
+        throw std::runtime_error("Failed to open file '" + filename +
+                                 "': " + err);
+    }
+    *handle = (void*)fd;
+}
+
+bool
+seek_and_write(void** handle, size_t offset, ConstByteSpan data)
+{
+    CHECK(handle);
+    auto* fd = reinterpret_cast<HANDLE*>(*handle);
+
+    auto* cur = reinterpret_cast<const char*>(data.data());
+    auto* end = cur + data.size();
+
+    int retries = 0;
+    OVERLAPPED overlapped = { 0 };
+    overlapped.hEvent = CreateEventA(nullptr, TRUE, FALSE, nullptr);
+
+    while (cur < end && retries < 3) {
+        DWORD written = 0;
+        DWORD remaining = (DWORD)(end - cur); // may truncate
+        overlapped.Pointer = (void*)offset;
+        if (!WriteFile(*fd, cur, (DWORD)remaining, nullptr, &overlapped) &&
+            GetLastError() != ERROR_IO_PENDING) {
+            const auto err = get_last_error_as_string();
+            LOG_ERROR("Failed to write to file: ", err);
+            CloseHandle(overlapped.hEvent);
+            return false;
+        }
+
+        if (!GetOverlappedResult(*fd, &overlapped, &written, TRUE)) {
+            LOG_ERROR("Failed to get overlapped result: ",
+                      get_last_error_as_string());
+            CloseHandle(overlapped.hEvent);
+            return false;
+        }
+        retries += (written == 0);
```

Comment:
Try to avoid implicit casting if possible.
```suggestion
        retries += (written == 0) ? 1 : 0;
```